packages/framework/src/Component/EntityExtension/EntityExtensionSubscriber.php

```diff
@@ -0,0 +1,306 @@
+<?php
+
+namespace Shopsys\FrameworkBundle\Component\EntityExtension;
+
+use Doctrine\Common\EventSubscriber;
+use Doctrine\ORM\Configuration;
+use Doctrine\ORM\Event\LoadClassMetadataEventArgs;
+use Doctrine\ORM\Events;
+use Doctrine\ORM\Mapping\ClassMetadataInfo;
+use LogicException;
+use ReflectionClass;
+use Webmozart\Assert\Assert;
+
+/**
+ * Inspired by joschi127/doctrine-entity-override-bundle (https://github.com/joschi127/doctrine-entity-override-bundle)
+ */
+class EntityExtensionSubscriber implements EventSubscriber
+{
+    /**
+     * @var array<class-string, class-string>
+     */
+    protected array $entityExtensionMap;
+
+    /**
+     * @var bool
+     */
+    protected bool $isEntityMapLoaded = false;
+
+    /**
+     * @var array<class-string, array<class-string, class-string>>
+     */
+    protected array $parentClassesByClass = [];
+
+    /**
+     * @var array<class-string, class-string>
+     */
+    protected array $allParentClasses = [];
+
+    /**
+     * @var class-string[]|null
+     */
+    protected ?array $allRegisteredEntitiesCache = null;
+
+    /**
+     * @var \Doctrine\ORM\Configuration
+     */
+    protected Configuration $configuration;
+
+    /**
+     * @param array<class-string, class-string> $entityExtensionMap
+     */
+    public function __construct(array $entityExtensionMap)
+    {
+        $this->setEntityExtensionMap($entityExtensionMap);
+    }
+
+    /**
+     * @return string[]
+     */
+    public function getSubscribedEvents(): array
+    {
+        return [Events::loadClassMetadata];
+    }
+
+    /**
+     * @param array<class-string, class-string> $entityExtensionMap
+     */
+    protected function setEntityExtensionMap(array $entityExtensionMap): void
+    {
+        $this->entityExtensionMap = $entityExtensionMap;
+        $this->isEntityMapLoaded = false;
```

Comment:
The method `setEntityExtensionMap` only sets the input data and the true load of the extension map is done in `loadEntityExtensionMap` method. 
Method `setEntityExtensionMap` is called also in tests in `OverwritableEntityExtensionMap`, and there we want to be sure that entity extension maps will be loaded again to prevent stale data, thus `false` here.

This all is done for performance reasons – the constructor is called upon each request, so it only set one property. The heavy lifting is done in method `loadClassMetadata` and this method is called only once.